Alternator: Support new ReturnValuesOnConditionCheckFailure feature

nuivall opened the pull request:
alternator: add support for ReturnValuesOnConditionCheckFailure feature

As announced in https://aws.amazon.com/about-aws/whats-new/2023/06/amazon-dynamodb-cost-failed-conditional-writes/, DynamoDB added a new option for write operations (PutItem, UpdateItem, or DeleteItem), ReturnValuesOnConditionCheckFailure, which if set to ALL_OLD returns the current value of the item - but only if a condition check failed.

Fixes https://github.com/scylladb/scylladb/issues/14481

nyh commented:
Please don't include the entire boto3 (!!) inside Scylla's repository, and if you really want to do this, do it in a separate PR, not inside a feature PR.

I took great care to ensure that test/alternator can run on all fairly-modern versions of boto3, and in the rare cases where the test requires a very recent boto3, the test is skipped if boto3 is too old - see examples in test_tag.py (where we need at least botocore 1.12.136 and if we don't have it, a test is skipped) and conftest.py (which insists that botocore is older than the antique 1.12.54, before which even BillingMode was not supported).

dbuild already includes a specific version of boto3, which will be used by Jenkins and other official tests scenarios, but developers should be able to use whatever version of boto3 they have installed, without getting yet another one of them inside our git repository.

nuivall commented:
> By the way, if you implement `ReturnValuesOnConditionCheckFailure` please also include in your patch a change to docs/alternator/compatibility.md, to remove the statement that it's not yet implemented. Thanks.

Thanks, removed.

nuivall commented:
> Please don't include the entire boto3 (!!) inside Scylla's repository, and if you really want to do this, do it in a separate PR, not inside a feature PR.
> 
> I took great care to ensure that test/alternator can run on all fairly-modern versions of boto3, and in the rare cases where the test requires a very recent boto3, the test is skipped if boto3 is too old - see examples in test_tag.py (where we need at least botocore 1.12.136 and if we don't have it, a test is skipped) and conftest.py (which insists that botocore is older than the antique 1.12.54, before which even BillingMode was not supported).
> 
> dbuild already includes a specific version of boto3, which will be used by Jenkins and other official tests scenarios, but developers should be able to use whatever version of boto3 they have installed, without getting yet another one of them inside our git repository.

I can retract checking it in no problem, I wasn't aware that jenkins runs tests in dbuild, it should be fine.

nyh commented:
Since this will be (after removing the boto3 mega-patch) a single-patch PR, please include "Fixes #14481" in the commit message itself. And also please instead of just a link, include some text describing the feature so someone browsing through git log or trying to build release notes won't need to read long documents and summarize them. You can perhaps copy-paste a paragraph from #14481 or paraphrase a couple of sentences on your own.

nyh commented on alternator/executor.cc:
```diff
@@ -1629,7 +1665,7 @@ future<executor::request_return_type> rmw_operation::execute(service::storage_pr
             db::consistency_level::LOCAL_SERIAL, db::consistency_level::LOCAL_QUORUM, timeout, timeout).then([this, read_command] (bool is_applied) mutable {
         if (!is_applied) {
-            return make_ready_future<executor::request_return_type>(api_error::conditional_check_failed("Failed condition."));
+            return make_ready_future<executor::request_return_type>(api_error::conditional_check_failed("The conditional request failed", std::move(_return_attributes)));
```
I see now where you used the new "JSON" parameter, but still don't understand how it reaches the user - isn't this _json field just ignored by the server? I don't see a patch to server.cc's generate_error_reply() which prints this error.
I see you changed the error message itself a bit - is it because the new message fits DynamoDB's message better?

nuivall replied:
yes, the message this is what dynamodb is currenty returning

nyh commented on alternator/executor.cc:
```diff
@@ -1716,8 +1752,7 @@ class put_item_operation : public rmw_operation {
     virtual std::optional<mutation> apply(std::unique_ptr<rjson::value> previous_item, api::timestamp_type ts) const override {
-        if (!verify_expected(_request, previous_item.get()) ||
-            !verify_condition_expression(_condition_expression, previous_item.get())) {
+        if (!verify_conditions(previous_item.get(), _condition_expression)) {
```
nitpick: please pick between verify_conditions() being a static function that takes all the parameters (like verify_expected() at al. were) or a member function that has access to all fields. Here you have this function take some fields (like _requests) from members, but the _condition_expression one as a parameter, which is weird.

I personally prefer the "static" approach because it makes it clearer exactly what each function does and needs.

nuivall replied:
It makes little sense to detach verify_conditions from its class by making it static, too many function parameters makes it less readable, class is for encapsulating logic including knowledge of what input parameters are, especially in this case where rmw_operation is encapsulating operation.

It's using _condition_expression as a parameter because every subclass of rmw_operation has a different field (with the same name). Maybe this should be fixed instead by moving the field to rmw_operation?

nyh replied:
Ok, I didn't remember/realize that _condition_expression is different in that it's not inherited but each has a separate variable with the same name.

I guess the OO vs static function thing is a matter of style, so feel free to keep your OO style. I personally dislike seeing related code get split to different header files and source files - a "static" function just resides in exactly the source file that needs it. I'm not bothered by a static function which takes 5 parameters - quite the opposite - it makes it clear that it needs exactly these 5 parameters, not less and not more.

nuivall replied:
I'm opting for keeping static usage to low, mainly because its widespread in alternator have bitten me a bit when developing expressions cache.

And in general while static functions and dependency injection seem to fulfill Open–closed principle and it's good to know dependencies and functions input in practice when codebase grows overusage leads to code which is very difficult to maintain and modify. As modification is what you need to do, can't really put more layers on layers every time.

nyh commented on test/alternator/test_returnvalues.py:
```diff
@@ -83,6 +83,28 @@ def test_put_item_returnvalues(test_table_s):
     with pytest.raises(ClientError, match='ValidationException'):
         test_table_s.put_item(Item={'p': p, 'a': 'hello'}, ReturnValues='none')
 
+# Testing feature which returns values only on delete condition failures.
```
Please explicitly use the feature's name, ReturnValuesOnConditionCheckFailure in the comment.  Same in all copies below. And please make the comments fit the test (here you said "delete" but it's a "put") and a bit clearer (it's not a "delete condition failure", it's a "DeleteItem with a failed condition expression").

nuivall replied:
ok

nyh commented on test/alternator/test_returnvalues.py:
```diff
@@ -83,6 +83,28 @@ def test_put_item_returnvalues(test_table_s):
+    except test_table_s.meta.client.exceptions.ConditionalCheckFailedException as err:
+        assert err.response['Item'] == {'p': {'S': p}, 's': {'S': 'cat'}}
+        got_exception = True
+    assert got_exception
```
nitpick: you can also use the more "idiomatic" pytest.raises, something like this (untested):
```python
with pytest.raises(ConditionalCheckFailedException) as err:
    test_table_s.put_item(...)
assert err.response['Item'] == ...
```
The raises idiom asserts that the exception must be thrown, and after the raises() check, you still have access to "err" and check that it's valid. So you don't need 3 lines dealing with "got_exception".

nuivall replied:
thanks, your approach works, I just needed to add err.value prefix to get to response object

nyh commented on test/alternator/test_returnvalues.py:
```diff
@@ -83,6 +83,28 @@ def test_put_item_returnvalues(test_table_s):
+        ConditionExpression='s = :v1',
+        ExpressionAttributeValues={':v1' : 'dog'})
+    assert not 'Attributes' in ret
+    # Failed conditional delete returns old values.
```
The following is conditional put, not conditional delete

nuivall replied:
thx

nyh commented on test/alternator/test_returnvalues.py:
```diff
@@ -441,3 +486,27 @@ def test_update_item_returnvalues_all_new_add_existing(test_table_s):
         ExpressionAttributeValues={':inc': 1})
     assert ret['Attributes']['n'] == 4
+
+# Testing feature which returns values only on modification condition failures.
```
Again all my comments above about this comment. In particular, it's not a "modification condition failure", is a "UpdateItem with a failed condition expression".

nuivall replied:
ok

nyh commented on test/alternator/test_returnvalues.py:
```diff
@@ -441,3 +486,27 @@ def test_update_item_returnvalues_all_new_add_existing(test_table_s):
+    except test_table_s.meta.client.exceptions.ConditionalCheckFailedException as err:
+        assert err.response['Item'] == {'p': {'S': p}, 's': {'S': 'cat'}}
+        got_exception=True
+    assert got_exception
```
These tests are missing the following cases which we also need to verify (and as usual check it behaves like DynamoDB):
1. When you use ReturnValuesOnConditionCheckFailure of None, it works but the item is not returned on failure.
2. When you use ReturnValuesOnConditionCheckFailure of "ALL_NEW" or "dog", you get an error message.

nuivall replied:
will add

nuivall commented on alternator/error.hh:
```diff
@@ -27,10 +30,13 @@ public:
         : _http_code(std::move(http_code))
         , _type(std::move(type))
         , _msg(std::move(msg))
+        , _json(std::move(json))
```
sorry, I accidentally stashed this change in server.cc, will add it.

nyh replied:
Ok, maybe after I'll see it I'll understand better. Did DynamoDB added to the protocol a way to return an item as part of an error, not just an error? Did you see it documented somewhere?

nuivall replied:
Yes, since it's still (condition) error this feature changes error class and dynamod does return non-200 code and adds  values json there. 

AFAICS it's not really documented I had to wireshark it to see that's inside "Item" object.

This can be moved to separate commit.

nyh replied:
Yes, please move it to a separate commit explaining the new layout of error messages. The original format documented in https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html had "__type" and "message" keys. What is the new key (I guess I'll see it when I see your change to server.cc)? Is there no trace of it in any DynamoDB documentation? That's so weird.

nyh replied:
I still can't find any better documentation, but I found this discussion: https://github.com/boto/boto3/issues/2657 on TransactionCanceledException, which suggests that in addition to the "__type" and "Message", the error message can have additional keys and values  - in that case they have the key "CancellationReasons" with an object as a value.

So perhaps we need in api_error the general ability to add any further keys and values, and in ConditionCheckException specifically add a parameter "item" (not "json") which sets the "Item" key in the error message. When we later implement transactions, we'll need to support TransactionCanceledException and this will set the "CancellationReasons" key - not the "Item" key.

So perhaps api::error() should just have a JSON map variable, and if it's not empty, the individual entries from this map will be inserted into the error message JSON?

nuivall commented:
v2:
- addressed comments
- added more tests
- removed boto3 from source
- moved error api change to separate commit (also made it more generic)

nyh commented on test/alternator/test_returnvalues.py:
```diff
@@ -83,6 +83,47 @@ def test_put_item_returnvalues(test_table_s):
+        test_table_s.put_item(Item={'p': p, 's': 'cow'},
+            ReturnValuesOnConditionCheckFailure='DUMMY',
+            ConditionExpression='s = :v1',
+            ExpressionAttributeValues={':v1' : 'dog'})
```
Thanks for adding all these tests.
I have another idea for a test: What if ReturnValues (not ReturnValuesOnConditionCheckFailure!) is set, and the condition failed - will it set the Item on the error or not?
If the answer is no, does combining ReturnValues and ReturnValuesOnConditionCheckFailure work? (it should return the item both in the successful case and error case).

nuivall replied:
this gets exponential I am afraid, since those features use different fields both for setup and output we can assume they are orthogonal, otherwise we won't ever stop adding new test cases

nyh replied:
One of the idea of making Alternator tests quick to run (right now we have 739 test functions taking 30 seconds sequentially), and quick to write, is that we don't need to be afraid to add new test cases.
I agree that we don't want to explore all combinations of different features - e.g., ReturnValuesOnConditionCheckFailure combined with a materialized view isn't particularly interesting. But when you as a developer know that two features are risky together because they touch the same code paths (e.g., ReturnValues and ReturnValuesOnConditionCheckFailure both mess with the same variable) or when a test checks some "if" case but doesn't check its "else" case, it can be useful to add more tests.
I don't insist on these additional tests - I agree that improving the test coverage is a never-ending effort and we can always add more later. But if you're looking for ideas on what you didn't test and might test - I wanted to suggest ideas.

nyh commented on test/alternator/test_returnvalues.py:
```diff
@@ -83,6 +83,47 @@ def test_put_item_returnvalues(test_table_s):
+    # Invalid ReturnValuesOnConditionCheckFailure value returns error.
+    with pytest.raises(ClientError, match='ValidationException'):
+        test_table_s.put_item(Item={'p': p, 's': 'cow'},
+            ReturnValuesOnConditionCheckFailure='DUMMY',
```
Thanks for adding this. Maybe you can also check (a loop?) a bunch of wrong names, including the names ALL_NEW etc. which are legal for ReturnValues but not for  ReturnValuesOnConditionCheckFailure.

nyh commented on test/alternator/test_returnvalues.py:
```diff
@@ -124,6 +165,41 @@ def test_delete_item_returnvalues(test_table_s):
         test_table_s.delete_item(Key={'p': p}, ReturnValues='none')
 
+# Testing ReturnValuesOnConditionCheckFailure feature which returns values only
+# on failed condition expression.
```
Nitpick - maybe mention the operation type in each of these identical comments - e.g., here say " ... on failed condition expression in DeleteItem operation", and in the previous test, it's the PutItem operation, and so on.

nuivall replied:
what would be the benefit of such addition?

nyh replied:
Just a better explanation of what this specific test does. How can several tests have exactly the same comment explaining what they do? :-)

nuivall replied:
I doubt anyone would find this out :) But my understanding is that primary description is entity's name (i.e. test name, function name, variable name) and comment provides additional info, it may even formulate the same thing using different wording - in case someone didn't understand. I see little value in having something like:

```
# Tests put item on error.
def test_put_item_on_error(...)
```

nyh replied:
The example you gave is irrelevant - both because in that example you really copied the function's name into a comment, which I'm not suggesting to do, and because you have a 5-word comment, which I also don't want you do - I want descriptions of tests to be multiple lines. You don't know how difficult it is to debug test failures when the test author didn't bother to explain why he thought this test is correct or relevant or what it's even supposed to test.

Anyway, your actual code is:
```
# Testing ReturnValuesOnConditionCheckFailure feature which returns values only
# on failed condition expression.
def test_delete_item_returnvalues_on_condition_check_failure(test_table_s):
```
Why do you want to "hide" the fact that this test is specific to DeleteItem? Yes, the name of the test has the words "delete_item". But it also has 6 other words. It's not impossible to figure out, but isn't it much easier just to say what this test actually does (and not have 3 different tests with exactly the same comment) instead of making it into a riddle? :-)

Anyway, I don't insist, I'll merge as-is.

nyh commented on test/alternator/test_returnvalues.py:
```diff
@@ -441,3 +517,41 @@ def test_update_item_returnvalues_all_new_add_existing(test_table_s):
+    skip_if_returnvalues_on_condition_check_not_supported()
+    p = random_string()
+    test_table_s.put_item(Item={'p': p, 's': 'dog'})
+    # Succesfull modification doesn't return values.
```
spelling: successful.

nyh commented on tools/toolchain/Dockerfile:
```diff
@@ -13,5 +13,8 @@ RUN dnf -y update \
 RUN mkdir -p /root/.m2/repository
+# we need at least 1.29.164 to run all alternator tests,
+# once on >= fedora 39 we can drop this as it has higer botocore by default
+RUN pip3 install --only-binary=:all: -v --upgrade botocore==1.29.164
```
What does "--only-binary" do?
What if we forget to remove this line in the future - will it be ignored if we already have higher botocore, or will it force this specific version?
Finally, I think our usual process isn't to edit this Dockerfile, but rather to edit `install_dependencies.sh`, and that contains the pip and other commands - so I think that's what you should do here. But I want to ask @avikivity to review this part of the patch.

I also want to comment that this change isn't strictly necessary: If we don't do it, it only means that the new test will be skipped. The test won't be skipped for users who manually installed a new botocore - so **you** will be able to test it, and hopefully we won't have a regession before Fedora 39 comes in with an updated botocore and this test returns to the regression testing regimen.

nuivall replied:
--only-binary should install wheel which is recommended way, I guess practical effect is that it will save some space.

--upgrade will force specific version, which should be fine

Putting it in install_dependencies.sh is a bit contradictory with previous requirement that test should use whatever library user has, as people are running this script frequently, but I can move it there if there is such request.

In general I'd like to have this test enabled on CI, as otherwise it could only cause troubles later (i.e. it would be run for the first time on CI when we switch to fedora 39).

I am guessing there is something additional to be done to upgrade image in docker hub after the patch is merged to make the Dockerfile change effective.

nuivall replied:
moved to install_dependencies.sh, also added >= requirement I hope should not downgrade anymore (I am still more in favor of pinning exact version, even exact source of dependencies but that's different discussion)

avikivity replied:
Please ask Fedora to update its botocore version, it's better not to carry too many special dependencies.

avikivity replied:
Exact version is problematic, if people move between branches it forces them to uninstall and reinstall.

There's always the frozen toolchain, but we like not to force it when we can.

nyh replied:
According to https://packages.fedoraproject.org/pkgs/python-botocore/python3-botocore/
Fedora 38 has 1.29.111, which supposedly is not new enough, and Fedora 39 will have 1.31.21.
I don't know how receptive they are to upgrading Python packages on Fedora 38, you can try.

But personally, I would do what I said before: Just make do with whichever botocore is installed. If it means that your new test will be skipped in Jenkins for the next 6 months, I think it's a small price to pay (this is why I asked to skip the test when botocore isn't new enough to perform it). You tested the new feature yourself on a newer botocore and it passed - it's unlikely that such a small feature will regress in 6 months, and moreover developers with newer botocore (like me) will test it sporadically. When we adopt Fedora 39 or Fedora 38 decides to upgrade botocore, the test will start running automatically.

nuivall replied:
ok, I will remove upgrade of botocore from install-dependecies

nyh reviewed (changes requested):
Thanks, looks great. I only left some minor comments (and one potential for crash?).

nuivall commented on test/alternator/test_returnvalues.py:
```diff
@@ -83,6 +83,47 @@ def test_put_item_returnvalues(test_table_s):
+    assert not 'Attributes' in ret # Ordinary field not present.
+    assert not 'Item' in ret # Field used for error responses also not present.
+    # Failed conditional put returns old values.
+    with pytest.raises(test_table_s.meta.client.exceptions.ConditionalCheckFailedException) as err:
```
looks like this is the only non string way (https://boto3.amazonaws.com/v1/documentation/api/latest/guide/error-handling.html), because aws error classes are dynamically generated, I am assuming they call some special endpoint to construct them or make them on the go as they receive error responses from AWS.

nuivall commented on test/alternator/test_returnvalues.py:
```diff
@@ -83,6 +83,47 @@ def test_put_item_returnvalues(test_table_s):
+            ReturnValuesOnConditionCheckFailure='ALL_OLD',
+            ConditionExpression='s = :v1',
+            ExpressionAttributeValues={':v1' : 'dog'})
+    assert err.value.response['Item'] == {'p': {'S': p}, 's': {'S': 'cat'}}
```
it was even difficult to find docs about the fact that response is put into 'Item' object

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3508/

nuivall commented on alternator/executor.cc:
```diff
@@ -1548,6 +1568,22 @@ std::optional<shard_id> rmw_operation::shard_for_execute(bool needs_read_before_
+        !verify_condition_expression(condition_expression, previous_item)) {
+        if (_returnvalues_on_condition_check_failure ==
+            returnvalues_on_condition_check_failure::ALL_OLD) {
+            _return_attributes = std::move(*previous_item);
```
check added

nyh replied:
Please also make sure you have a test (or part of a test) exercising this new check. This is much easier to do while you are developing the code, and not later. And the the test will also serve as a regression test (if someone ever rewrites this code and forgets to handle this case, the test will catch it).

nuivall replied:
added

nuivall commented:
v3:

- more tests
- check for null ptr added
- added wrapper class for copying value
- removed shared pointer, mainly because api_error can be deleted on different shard than constructed.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4001/

nuivall commented:
v4: removed install-dependencies botocore upgrade

btw jenkins job failed but doesn't look like anything related:
```
17:23:51  Error: Could not publish |scylla/build/debug/dist/tar/scylla-*x86_64.tar.gz| from dir |/jenkins/workspace/scylla-master/scylla-ci| to S3 bucket |downloads.scylladb.com/unstable/scylla/master/relocatable/2023-09-28T18:06:45Z| Second try. Error: |java.io.IOException: Failed to upload files|
```

nyh commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4001/

Restarted the CI

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4004/

nuivall commented on utils/rjson.hh:
```diff
@@ -165,6 +165,27 @@ inline std::string_view to_string_view(const rjson::value& v) {
+    copyable_value(value&& v) noexcept : value(std::move(v)) {
+    }
+    copyable_value& operator=(const copyable_value& v) {
+        *this = copy(v);
```
added

nuivall commented:
v5: 
added self assignment protection

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4010/

nuivall commented on alternator/executor.cc:
```diff
@@ -2796,9 +2830,8 @@ static bool hierarchy_actions(
-        !verify_condition_expression(_condition_expression, previous_item.get())) {
-        // If the update is to be cancelled because of an unfulfilled
+    if (!verify_conditions(previous_item.get(), _condition_expression)) {
+        // If the update is to be cancelled because of an unfulfilled Expected
```
reverted

nuivall commented on alternator/executor.cc:
```diff
@@ -1548,6 +1568,22 @@ std::optional<shard_id> rmw_operation::shard_for_execute(bool needs_read_before_
+// verify_conditions checks if previous (current) value of the
+// item matches the conditions (Expected or ConditionExpression) for modification operation.
+// Additionally when ReturnValuesOnConditionCheckFailure field in the request is set
+// it moves previous_item so that it can be returned in the response.
```
reverted back to static

nuivall commented:
v6:
- reverted back to static functions for condition checks

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - Build
:x: - Sanity Tests
:x: - Unit Tests
#### Failed Tests (2/21047):
- [test_increment_decrement_counters_in_threads_nodes_restarted](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4439/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___Sanity_Tests___test_increment_decrement_counters_in_threads_nodes_restarted) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_increment_decrement_counters_in_threads_nodes_restarted)
- [test_exception_safety_of_reads](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4439/testReport/junit/boost/row_cache_test/test_exception_safety_of_reads) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_exception_safety_of_reads)
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4439/
- Duration: 4 hr 29 min
- Builder: i-072458e1c0a8d080c

nuivall commented:
> Failed Tests (2/21047):
[test_increment_decrement_counters_in_threads_nodes_restarted](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4439/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___Sanity_Tests___test_increment_decrement_counters_in_threads_nodes_restarted) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_increment_decrement_counters_in_threads_nodes_restarted)
[test_exception_safety_of_reads](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4439/testReport/junit/boost/row_cache_test/test_exception_safety_of_reads) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_exception_safety_of_reads)

Doesn't look related to alternator at all.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4554/
- Duration: 2 hr 9 min
- Builder: spider4.cloudius-systems.com

nuivall commented:
@nyh ping
